vickyctl/src/locks.rs

```diff
@@ -1,30 +1,433 @@
+                    if selected_task.is_some() {
+                        *selected_task = None;
+                        return Ok(false);
+                    }
+                    return Ok(true);
+                }
+
+                match selected_task {
+                    None => handle_task_list(state, lock_amount, selected_task, &key),
+                    Some(_) => handle_popup(selected_task, selected_button, &key, args, locks)?,
+                }
+            }
+        }
+    }
+    Ok(false)
+}
+
+fn handle_task_list(
+    state: &mut TableState,
+    lock_amount: usize,
+    selected_task: &mut Option<usize>,
+    key: &KeyEvent,
+) {
+    match state.selected_mut() {
+        None => (),
+        Some(cur) => {
+            if (key.code == KeyCode::Up || key.code == KeyCode::Char('k')) && *cur > 0 {
+                *cur -= 1;
+            } else if (key.code == KeyCode::Down || key.code == KeyCode::Char('j'))
+                && *cur < lock_amount - 1
+            {
+                *cur += 1;
+            } else if key.code == KeyCode::Enter {
+                *selected_task = Some(*cur);
+            }
+        }
+    };
+}
+
+#[allow(dead_code)]
+fn get_longest_len<'a, T>(str_iter: T) -> u16
+where
+    T: Iterator<Item = &'a str>,
+{
+    str_iter
+        .map(|l| l.len())
+        .max()
+        .map_or(0, |len| u16::try_from(len).unwrap_or(u16::MAX))
+}
+
+fn draw_task_picker(f: &mut Frame, locks: &[PoisonedLock], state: &mut TableState) {
+    let rows: Vec<Row> = locks.iter().map(|l| l.into()).collect();
+
+    // let _widths: [Constraint; 4] = [
+    //     Constraint::Max(get_longest_len(locks.iter().map(|l| l.name()))),
+    //     Constraint::Max(5),
+    //     Constraint::Max(get_longest_len(
+    //         locks
+    //             .iter()
+    //             .map(|l| l.get_poisoned_by().display_name.as_str()),
+    //     ).max("Failed Task Name".len() as u16)),
+    //     Constraint::Min(get_longest_len(
+    //         locks
+    //             .iter()
+    //             .map(|l| l.get_poisoned_by().flake_ref.flake.as_str()),
+    //     ).max("Task Flake URI".len() as u16)),
+    // ]; // This will not make the table equally spaced, but instead use minimal space. Just keeping this since I don't wanna rewrite it.
+    let table = Table::new(rows, &[])
+        .block(
+            Block::default()
+                .title("Manually Resolve Locks")
+                .title_alignment(Alignment::Center)
+                .borders(Borders::ALL),
+        )
+        .header(
+            Row::new(vec!["Name", "Type", "Failed Task Name", "Task Flake URI"])
+                .set_style(Style::default().bold().italic()),
+        )
+        .highlight_symbol(">>")
+        .highlight_style(Style::default().fg(Color::Green).italic())
+        .highlight_spacing(HighlightSpacing::Always);
+    f.render_stateful_widget(table, f.size(), state);
+}
+
+fn draw_confirm_clear(
```

Comment:
We should extract those functions from here into a form of framework for reusability and uniformity.

Reply:
the popup functions?

Reply:
Yeah, any tui-related rendering functions.